common/oidc-client/oauth2-proxy/components/istio-m2m/requestauthentication.yaml

```diff
@@ -0,0 +1,14 @@
+apiVersion: security.istio.io/v1beta1
+kind: RequestAuthentication
+metadata:
+  name: m2m-token-issuer
+  namespace: istio-system
+spec:
+  jwtRules:
+    - forwardOriginalToken: true
+      issuer: M2M_TOKEN_ISSUER_PLACEHOLDER
+      outputClaimToHeaders:
+        - claim: sub
+          header: x-auth-request-user
+        - claim: sub
```

Comment:
i asume that
```
        - header: kubeflow-userid
          claim: sub
```

means use the claim "sub" from the JWT (serviceaccountoken) and use it as kubeflow-userid header. This header is then used in most of our authorizationpolicies https://github.com/kimwnasptd/kubeflow/blob/52d55bef092d9ec56a1b4c6ae8d57fff241ce2fe/components/access-management/kfam/bindings.go#L95 and also in most of our webservices

Sub is typically "system:serviceaccount:kubeflowusernamespace:default-editor"
while in the DEX JWT we use the users email to set the header kubeflow-userid.
Now i am wondering how our authorizationpolicies are going to react to "system:serviceaccount:kubeflowusernamespace:default-editor" as userid header instead of the email.

Reply:
Yes, I totally agree on the headers and bindings nightmare.

> use the claim "sub" from the JWT (serviceaccountoken) and use it as kubeflow-userid header

That's exactly true.

> how our authorizationpolicies are going to react to "system:serviceaccount:kubeflowusernamespace:default-editor" as userid header instead of the email

Well, technically with the m2m sa tokens you're not reaching the kubeflow user namespace directly but the KF API which has means of authorization through the `TokenReviews` and `SubjectAccessReviews`.

Let's analyze the the code in a following scenario:
1. api-server starts endpoints in

     https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/main.go#L95

    Let's focus on the pipelines run service:
    1. Register Run Service:
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/main.go#L116
    2. proto RPC definition of ListRunsV1
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/api/v1beta1/run.proto#L80
    3. code definition of ListRunsV1
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/server/run_server.go#L226
    4. ListRunsV1 calls internal method `listRuns`
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/server/run_server.go#L188 
    5. `listRuns` calls internal method `canAccessRun` which itself calls `s.resourceManager.IsAuthorized`
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/server/run_server.go#L637
    6. `ResourceManager.IsAuthorized` first tries to authenticate over every available authenticator, which are the `TokenReviewAuthenticator` and `HTTPHeaderAuthenticator`
        * here the user identity is either the user email provided directly in the `kubeflow-userid` header or the user identity obtained from provided token
        * https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/resource/resource_manager.go#L1667
    8. `TokenReviewAuthenticator.GetUserIdentity` gets the token from `Authorization` header and calls the K8s Auth `authv1.TokenReview` with given token which in return provides `userInfo := review.Status.User`. `GetUserIdentity` return `userInfo.Username` which at this point is the `system:serviceaccount:default:default`.
        * https://github.com/kubeflow/pipelines/blob/2.0.5/backend/src/apiserver/auth/authenticator_token_review.go#L53
    9. Next in `ResourceManager.IsAuthorized` a SubjectAccessReview is created with `r.subjectAccessReviewClient.Create` with arguments specifying RBAC verbs provided in code definition of `RunServer.listRuns`. If the user (sa) is not authorized, an error is thrown
        * https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/resource/resource_manager.go#L1703
        * if the identity was obtained from token (service account), the `rolebinding.rbac.authorization.k8s.io/default-editor` provides the RBAC permission
        * if the identity was obtained from header (user), the `rolebinding.rbac.authorization.k8s.io/user-example-com` or similar provides the RBAC permission

1. User calls api to list pipeline runs as unauthorized service account.
    * This can be done by running Pod with curl in `default` namespace:
        ```bash
        $ kubectl -n default run -ti --rm curl --image curlimages/curl --command -- sh
        # v1beta1
        ~ $ curl "istio-ingressgateway.istio-system/pipeline/apis/v1beta1/runs?resource_reference_key.type=NAMESPACE&resource_reference_key.id=kubeflow-user-example-com" -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)"
        {"error":"Failed to list v1beta1 runs: Failed to list runs due to authorization error. Check if you have permission to access namespace kubeflow-user-example-com: Failed to access run . Check if you have access to namespace kubeflow-user-example-com: PermissionDenied: User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,}): Unauthorized access","code":7,"message":"Failed to list v1beta1 runs: Failed to list runs due to authorization error. Check if you have permission to access namespace kubeflow-user-example-com: Failed to access run . Check if you have access to namespace kubeflow-user-example-com: PermissionDenied: User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,}): Unauthorized access","details":[{"@type":"type.googleapis.com/google.rpc.Status","code":7,"message":"User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,})"}]}
        # v2beta1
        ~ $ curl istio-ingressgateway.istio-system/pipeline/apis/v2beta1/runs?namespace=kubeflow-user-example-com -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)"
        {"error":"Failed to list runs: Failed to list runs due to authorization error. Check if you have permission to access namespace kubeflow-user-example-com: Failed to access run . Check if you have access to namespace kubeflow-user-example-com: PermissionDenied: User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,}): Unauthorized access","code":7,"message":"Failed to list runs: Failed to list runs due to authorization error. Check if you have permission to access namespace kubeflow-user-example-com: Failed to access run . Check if you have access to namespace kubeflow-user-example-com: PermissionDenied: User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,}): Unauthorized access","details":[{"@type":"type.googleapis.com/google.rpc.Status","code":7,"message":"User 'system:serviceaccount:default:default' is not authorized with reason:  (request: \u0026ResourceAttributes{Namespace:kubeflow-user-example-com,Verb:list,Group:pipelines.kubeflow.org,Version:v1beta1,Resource:runs,Subresource:,Name:,})"}]}
        ```

1. User calls api to list pipeline runs as authorized service account.
    * This can be done by running Pod with curl in `kubeflow-user-example-com` namespace specifying correct service account:
        ```bash
        $ kubectl -n kubeflow-user-example-com run -ti --rm curl --image curlimages/curl --command --overrides='{"spec": {"serviceAccountName": "default-editor"}}' -- sh
        # v1beta1
        ~ $ curl "istio-ingressgateway.istio-system/pipeline/apis/v1beta1/runs?resource_reference_key.type=NAMESPACE&resource_reference_key.id=kubeflow-user-example-com" -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)"
        {}  # empty response which is fine because no pipeline runs exist
        # v2beta1
        ~ $ curl istio-ingressgateway.istio-system/pipeline/apis/v2beta1/runs?namespace=kubeflow-user-example-com -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)"
        {}  # empty response which is fine because no pipeline runs exist
        ```

Having that said, I don't think we have to tailor the RoleBindings since the RBAC is already properly configured.

The AuthorizationPolicies might require a little tailoring for kserve... I wonder what are the scenarios when we want to access the Pod/SVC run in a given Namespace directly.

Reply:
I have to read this tomorrow, but pipelines is a bad example because it has special stuff for handling service account tokens. I am more interested in Notebooks, Kserve. Can you provide this analysis for them as well?

Reply:
We really need test for kserve, notebooks, pipelines automation via m2m tokens

Reply:
## Simplified analysis for Notebooks:

* list notebooks api route definition
  * https://github.com/kubeflow/kubeflow/blob/v1.8.0/components/crud-web-apps/jupyter/backend/apps/common/routes/get.py#L53
  * this calls `crud_backend/api/notebook.py::list_notebooks`
* `crud_backend/api/notebook.py::list_notebooks` calls `authz.ensure_authorized`
  * https://github.com/kubeflow/kubeflow/blob/v1.8.0/components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/api/notebook.py#L24
* `crud_backend/authz.py::ensure_authorized` calls `crud_backend/authn.py::get_username`
  * https://github.com/kubeflow/kubeflow/blob/v1.8.0/components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/authz.py#L101
  * https://github.com/kubeflow/kubeflow/blob/v1.8.0/components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/authn.py#L12
  * `crud_backend/authn.py::get_username` gets the user id from userid header (email or sa in format `system:serviceaccount:kubeflowusernamespace:default-editor`)
* `crud_backend/authz.py::ensure_authorized` calls `crud_backend/authz.py::is_authorized`
  * https://github.com/kubeflow/kubeflow/blob/v1.8.0/components/crud-web-apps/common/backend/kubeflow/kubeflow/crud_backend/authz.py#L46
  * this calls `create_subject_access_review` which uses the same mechanism as pipelines with `r.subjectAccessReviewClient.Create`

### Summary
Notebooks api uses the same mechanism for user authentication, proxying auth decision to K8s RBAC. 
* if the identity was obtained from token (service account), the `rolebinding.rbac.authorization.k8s.io/default-editor` provides the RBAC permission
* if the identity was obtained from header (user), the `rolebinding.rbac.authorization.k8s.io/user-example-com` or similar provides the RBAC permission

Reply:
On kserve I found this information:
* https://github.com/kserve/kserve/blob/v0.12.0/config/default/manager_auth_proxy_patch.yaml

This suggests the kserve **might** use the same mechanism based on `SubjectAccessReviews`. Having a look at the kubeflow/manifests I see it's not enabled.

Also, looking through the docs and code I was only able to find this:
* https://github.com/kserve/kserve/tree/v0.12.0/docs/samples/istio-dex
* https://github.com/kserve/kserve/tree/v0.12.0/docs/samples/gcp-iap

The docs above mention that while it's possible to enable authentication, authorization is more complicated and probably we need to add `AuthorizationPolicy`...
> create an [Istio AuthorizationPolicy](https://istio.io/latest/docs/reference/config/security/authorization-policy/) to grant access to the pods or disable it

I executed an experiment on example kubeflow installation to verify access from different namespace to kserve inferenceservice deployed in `kubeflow-user-example-com`:
1. Created InferenceService with sklearn
    ```yaml
    apiVersion: serving.kserve.io/v1beta1
    kind: InferenceService
    metadata:
      name: isvc-sklearn
      namespace: kubeflow-user-example-com
    spec:
      predictor:
        minReplicas: 1
        model:
          modelFormat:
            name: sklearn
          name: ''
          resources:
            limits:
              cpu: 100m
              memory: 256Mi
            requests:
              cpu: 50m
              memory: 128Mi
          storageUri: gs://kfserving-examples/models/sklearn/1.0/model
    ```
2. Created AuthorizationPolicy for given InferenceService Workload:
    ```yaml
    apiVersion: security.istio.io/v1
    kind: AuthorizationPolicy
    metadata:
      labels:
        app: isvc-sklearn-predictor-00001
      name: isvc-sklearn-predictor-00001
      namespace: kubeflow-user-example-com
    spec:
      action: DENY
      rules:
      - {}
      selector:
        matchLabels:
          app: isvc-sklearn-predictor-00001
    ```
3. Called inference service from a default namespace using `default:default` SA Token:
    ```bash
    $ kubectl -n default run -ti --rm curl --image curlimages/curl --command -- sh
    ~$ curl istio-ingressgateway.istio-system/v1/models/isvc-sklearn:predict -H "Authorization: Bearer $(cat /run/secrets/kubernetes.io/serviceaccount/token)" -H "Host: isvc-sklearn.kserve-test.example.com" -H "Content-Type: application/json" -d '{"instances": [[6.8, 2.8, 4.8, 1.4], [6.0, 3.4, 4.5, 1.6]]}'
    {"predictions":[1,1]}
    ```

I'm confused by why this AuthorizationPolicy didnt work. But, looking at the documentation I noticed that the [kserve 0.11.0 docs with istio-dex sample](https://github.com/kserve/kserve/tree/v0.11.0/docs/samples/istio-dex) doesn't mention AuthorizationPolicy but [kserve 0.11.0 docs with istio-dex sample](https://github.com/kserve/kserve/tree/v0.12.0/docs/samples/istio-dex) does. Maybe it's only enabled from future version...? I'm not sure and I'd welcome some comments or ideas.

Reply:
@kromanow94 For the InferenceService AuthorizationPolicies, you shoudn't match against the predictor revision number `app: isvc-sklearn-predictor-00001`, this will be inconsistent with the actual request flow as there can be multiple revisions for this given InfService.
Better use some other generally available label on the Deployments like `component=predictor` which is added to all the Deployments managed by this InferenceService.

Reply:
Hey @rawc0der , thanks for the suggestion! I'll experiment with it.